File: Algorithm/LeetCode/daily/2024_09/18_2332_mid.py
from bisect import bisect_left, bisect_right, insort_left, insort_right, insort, bisect
from math import ceil, floor, pow, gcd, sqrt, log10, fabs, fmod, factorial, inf, pi, e
from heapq import heapify, heapreplace, heappush, heappop, heappushpop, nlargest, nsmallest
from collections import defaultdict, Counter, deque
from itertools import permutations, combinations, combinations_with_replacement, accumulate, count, groupby, pairwise
from queue import PriorityQueue, Queue, LifoQueue
from functools import lru_cache, cache, reduce
from typing import List, Optional
import sys

from sortedcontainers import SortedList, SortedDict, SortedSet

sys.setrecursionlimit(10001000)

MOD = int(1e9 + 7)
INF = int(1e20)
INFMIN = float('-inf')  # 负无穷
INFMAX = float('inf')  # 正无穷
PI = 3.141592653589793
direc = [(1, 0), (0, 1), (-1, 0), (0, -1)]
direc8 = [(1, 0), (0, 1), (-1, 0), (0, -1), (1, 1), (1, -1), (-1, 1), (-1, -1)]
ALPS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
alps = 'abcdefghijklmnopqrstuvwxyz'

# @Author  : https://github.com/hakusai22
# @Time    : 2024/09/19 22:41
# @题目     : https://leetcode.cn/problems/the-latest-time-to-catch-a-bus/description/?envType=daily-question&envId=2024-09-18
# @参考     : https://leetcode.cn/problems/the-latest-time-to-catch-a-bus/solutions/2920715/python3javacgotypescript-yi-ti-yi-jie-mo-dgzf/?envType=daily-question&envId=2024-09-18
# 时间复杂度 : O(n×logn+m×logm)


class Solution:
    def latestTimeCatchTheBus(self, buses: List[int], passengers: List[int], capacity: int) -> int:
        global c
        buses.sort()
        passengers.sort()
        j = 0
        for t in buses:
            c = capacity
            while c and j < len(passengers) and passengers[j] <= t:
                c, j = c - 1, j + 1
        j -= 1
        # 若有空位，我们可以在公交车发车时 bus[∣bus∣−1] 到达公交站
        ans = buses[-1] if c else passengers[j]
        # 若无空位，我们可以找到上一个上车的乘客，顺着他往前找到没人到达的时刻。
        while j >= 0 and passengers[j] == ans:
            ans, j = ans - 1, j - 1
        return ans

File: Algorithm/LeetCode/daily/2024_09/18_2332_mid.go
package _024_09

import "sort"

// latestTimeCatchTheBus 计算最晚登车时间以抓住巴士。
// 参数:
//
//	buses []int: 巴士发车时间数组。
//	passengers []int: 乘客到达时间数组。
//	capacity int: 巴士容量。
//
// 返回值:
//
//	int: 乘客可以最晚登车的时间。
//
// 说明:
//
//	该函数通过排序巴士和乘客数组，然后模拟每个巴士的乘客搭载过程来找到最晚登车时间。
//	它考虑了巴士容量限制，并在达到巴士时间之前尽可能多的乘客。
//	时间复杂度: O(n log n)，其中 n 是巴士或乘客的数量。
//	空间复杂度: O(1)，使用了额外常数空间。
func latestTimeCatchTheBus(buses []int, passengers []int, capacity int) int {
	// 对巴士和乘客时间进行排序，以便按时间顺序处理。
	sort.Ints(buses)
	sort.Ints(passengers)

	// j 用于追踪当前处理的乘客索引，c 用于记录当前巴士剩余容量。
	j, c := 0, 0

	// 遍历每个巴士的发车时间。
	for _, t := range buses {
		// 重置当前巴士的容量。
		c = capacity

		// 当巴士还有空位，且还有未处理的乘客，且下一个乘客能在当前巴士发车前上车时，让乘客上车。
		for c > 0 && j < len(passengers) && passengers[j] <= t {
			j++
			c--
		}
	}

	// j 减一，因为我们可能在最后一个巴士发车时间之后没有乘客，或者最后一个乘客已经在巴士上。
	j--

	// 初始化答案为最后一个巴士的发车时间，这是理论上最晚可以登车的时间。
	ans := buses[len(buses)-1]

	// 如果最后一个巴士满员了，那么最晚登车时间是最后一个上车乘客的时间。
	if c == 0 {
		ans = passengers[j]
	}

	// 如果最晚登车时间与任何乘客到达时间相同，我们需要回溯到前一个时间，以确保不会与任何乘客冲突。
	for j >= 0 && ans == passengers[j] {
		ans--
		j--
	}

	// 返回最终计算出的最晚登车时间。
	return ans
}
